File: src/lib/ollama.ts
/* ──────────────────────────────────────────────────────────
   Orchestra AI — Ollama HTTP Client
   Communicates with the local Ollama server for inference
   ────────────────────────────────────────────────────────── */

import type { OllamaModel } from "@/types";

const DEFAULT_BASE_URL = "http://localhost:11434";

export interface OllamaGenerateRequest {
  model: string;
  prompt: string;
  system?: string;
  images?: string[]; // base64-encoded
  stream?: boolean;
  options?: {
    temperature?: number;
    num_predict?: number;
    top_p?: number;
    top_k?: number;
  };
}

export interface OllamaGenerateResponse {
  model: string;
  response: string;
  done: boolean;
  total_duration?: number;
  load_duration?: number;
  eval_count?: number;
  eval_duration?: number;
}

export interface OllmaChatMessage {
  role: "system" | "user" | "assistant";
  content: string;
  images?: string[];
}

export interface OllamaChatRequest {
  model: string;
  messages: OllmaChatMessage[];
  stream?: boolean;
  options?: {
    temperature?: number;
    num_predict?: number;
  };
}

class OllamaClient {
  private baseUrl: string;

  constructor(baseUrl?: string) {
    this.baseUrl = baseUrl || DEFAULT_BASE_URL;
  }

  setBaseUrl(url: string) {
    this.baseUrl = url;
  }

  /** Check if Ollama is running */
  async ping(): Promise<boolean> {
    try {
      const res = await fetch(`${this.baseUrl}/api/version`, {
        signal: AbortSignal.timeout(3000),
      });
      return res.ok;
    } catch {
      return false;
    }
  }

  /** List available models */
  async listModels(): Promise<OllamaModel[]> {
    try {
      const res = await fetch(`${this.baseUrl}/api/tags`);
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
      const data = await res.json();
      return data.models || [];
    } catch (err) {
      console.error("[Ollama] Failed to list models:", err);
      return [];
    }
  }

  /** Generate a completion (non-streaming) */
  async generate(req: OllamaGenerateRequest): Promise<OllamaGenerateResponse> {
    const res = await fetch(`${this.baseUrl}/api/generate`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ ...req, stream: false }),
    });
    if (!res.ok) {
      const text = await res.text();
      throw new Error(`Ollama generate failed (${res.status}): ${text}`);
    }
    return res.json();
  }

  /** Chat completion (non-streaming) */
  async chat(req: OllamaChatRequest): Promise<OllamaGenerateResponse> {
    const res = await fetch(`${this.baseUrl}/api/chat`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ ...req, stream: false }),
    });
    if (!res.ok) {
      const text = await res.text();
      throw new Error(`Ollama chat failed (${res.status}): ${text}`);
    }
    const data = await res.json();
    return {
      model: data.model,
      response: data.message?.content || "",
      done: data.done,
      total_duration: data.total_duration,
      eval_count: data.eval_count,
      eval_duration: data.eval_duration,
    };
  }

  /** Generate with streaming — yields chunks */
  async *generateStream(
    req: OllamaGenerateRequest
  ): AsyncGenerator<string, void, unknown> {
    const res = await fetch(`${this.baseUrl}/api/generate`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ ...req, stream: true }),
    });
    if (!res.ok || !res.body) {
      throw new Error(`Ollama stream failed (${res.status})`);
    }
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";
      for (const line of lines) {
        if (!line.trim()) continue;
        try {
          const json = JSON.parse(line);
          if (json.response) yield json.response;
        } catch {
          // skip malformed lines
        }
      }
    }
  }

  /** Pull a model (initiate download) */
  async pullModel(
    name: string,
    onProgress?: (status: string, completed?: number, total?: number) => void
  ): Promise<void> {
    const res = await fetch(`${this.baseUrl}/api/pull`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ name, stream: true }),
    });
    if (!res.ok || !res.body) {
      throw new Error(`Failed to pull model ${name}`);
    }
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = "";

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split("\n");
      buffer = lines.pop() || "";
      for (const line of lines) {
        if (!line.trim()) continue;
        try {
          const json = JSON.parse(line);
          onProgress?.(json.status, json.completed, json.total);
        } catch {
          // skip
        }
      }
    }
  }

  /** Vision completion: analyze an image */
  async analyzeImage(
    model: string,
    prompt: string,
    imageBase64: string,
    systemPrompt?: string
  ): Promise<string> {
    const result = await this.generate({
      model,
      prompt,
      system: systemPrompt,
      images: [imageBase64],
      options: { temperature: 0.4, num_predict: 2048 },
    });
    return result.response;
  }
}

/** Singleton Ollama client */
export const ollama = new OllamaClient();
export default OllamaClient;

File: src-tauri/src/ollama.rs
// Orchestra AI — Ollama HTTP Client (Rust)

use reqwest::Client;
use serde::{Deserialize, Serialize};

const DEFAULT_OLLAMA_URL: &str = "http://localhost:11434";

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct OllamaModel {
    pub name: String,
    pub size: u64,
    pub digest: String,
    #[serde(rename = "modified_at")]
    pub modified_at: String,
    pub details: Option<ModelDetails>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ModelDetails {
    pub format: Option<String>,
    pub family: Option<String>,
    #[serde(rename = "parameter_size")]
    pub parameter_size: Option<String>,
    #[serde(rename = "quantization_level")]
    pub quantization_level: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct GenerateRequest {
    pub model: String,
    pub prompt: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub system: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub images: Option<Vec<String>>,
    pub stream: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub options: Option<GenerateOptions>,
}

#[derive(Debug, Serialize)]
pub struct GenerateOptions {
    pub temperature: f32,
    pub num_predict: u32,
}

#[derive(Debug, Deserialize)]
pub struct GenerateResponse {
    pub model: String,
    pub response: String,
    pub done: bool,
    pub total_duration: Option<u64>,
    pub eval_count: Option<u32>,
    pub eval_duration: Option<u64>,
}

#[derive(Debug, Deserialize)]
struct TagsResponse {
    models: Vec<OllamaModel>,
}

pub struct OllamaClient {
    client: Client,
    base_url: String,
}

impl OllamaClient {
    pub fn new() -> Self {
        Self {
            client: Client::new(),
            base_url: DEFAULT_OLLAMA_URL.to_string(),
        }
    }

    pub async fn ping(&self) -> bool {
        self.client
            .get(format!("{}/api/version", self.base_url))
            .timeout(std::time::Duration::from_secs(3))
            .send()
            .await
            .map(|r| r.status().is_success())
            .unwrap_or(false)
    }

    pub async fn list_models(&self) -> Result<Vec<OllamaModel>, String> {
        let resp = self.client
            .get(format!("{}/api/tags", self.base_url))
            .send()
            .await
            .map_err(|e| format!("Failed to connect: {}", e))?;

        let tags: TagsResponse = resp
            .json()
            .await
            .map_err(|e| format!("Failed to parse: {}", e))?;

        Ok(tags.models)
    }

    pub async fn generate(&self, req: GenerateRequest) -> Result<GenerateResponse, String> {
        let resp = self.client
            .post(format!("{}/api/generate", self.base_url))
            .json(&req)
            .send()
            .await
            .map_err(|e| format!("Request failed: {}", e))?;

        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().await.unwrap_or_default();
            return Err(format!("Ollama error ({}): {}", status, text));
        }

        resp.json()
            .await
            .map_err(|e| format!("Parse error: {}", e))
    }
}

impl Default for OllamaClient {
    fn default() -> Self {
        Self::new()
    }
}
